File: implementations/typescript/packages/core/src/crypto.ts
// SPDX-License-Identifier: Apache-2.0
/** Node cryptography adapter. Codec/root entry point remains free of Node imports. */
import { createHmac, createPrivateKey, createPublicKey, sign, verify, timingSafeEqual } from "node:crypto";
import { PspError, record, validateJson } from "./json.js";
import { decodeSignature, encodeSignature, envelopeFromObject, signatureInput, validateTime, type Envelope, type Signature } from "./signatures.js";
import { documentToObject, type Document } from "./markup.js";

export interface TrustedKey {
  id: string; algorithm: Signature["algorithm"]; material: Uint8Array; status: "active" | "revoked";
  trustLevels: readonly number[]; sectionTypes: readonly string[]; scope: Record<string, string>; allowUnscoped: boolean;
}
export interface VerificationPolicy { keys: readonly TrustedKey[]; now: number; context: Record<string, string>; allowedAttributes: readonly string[]; clockSkew?: number }
function edKey(raw: Uint8Array, privateKey: boolean) {
  if (!(raw instanceof Uint8Array) || raw.length !== 32) throw new PspError("INVALID_KEY");
  const bytes = Buffer.concat([Buffer.from(privateKey ? "302e020100300506032b657004220420" : "302a300506032b6570032100", "hex"), raw]);
  return privateKey ? createPrivateKey({key:bytes, format:"der", type:"pkcs8"}) : createPublicKey({key:bytes, format:"der", type:"spki"});
}
function secret(raw: Uint8Array): Uint8Array {
  if (!(raw instanceof Uint8Array) || raw.length < 32) throw new PspError("INVALID_KEY", "HMAC keys must contain at least 32 bytes.");
  return raw;
}
export function signEnvelope(data: unknown, metadata: Omit<Signature, "value">, privateKey: Uint8Array): Envelope {
  const value = encodeSignature(new Uint8Array(metadata.algorithm === "ed25519" ? 64 : 32));
  const e = envelopeFromObject({signature:{...metadata, value}, data});
  if (validateTime(e.signature.timestamp, e.signature.expires, e.signature.timestamp) !== "valid") throw new PspError("INVALID_TIME");
  const bytes = signatureInput(e);
  e.signature.value = encodeSignature(e.signature.algorithm === "ed25519" ? sign(null, bytes, edKey(privateKey, true)) : createHmac("sha256", secret(privateKey)).update(bytes).digest());
  return e;
}
export function signDocument(document:Document, metadata:Omit<Signature,"value"|"contentType">, privateKey:Uint8Array, preserveSource=true):Envelope {
  return signEnvelope(documentToObject(document,preserveSource),{...metadata,contentType:"json"},privateKey);
}
/** Crypto only. Use verifyEnvelope to additionally require key authority, scope and time. */
export function verifySignature(input: Envelope, publicKeyOrSecret: Uint8Array): boolean {
  const e = envelopeFromObject(input), bytes = signatureInput(e), signature = decodeSignature(e.signature.value, e.signature.algorithm);
  return e.signature.algorithm === "ed25519" ? verify(null, bytes, edKey(publicKeyOrSecret, false), signature)
    : timingSafeEqual(createHmac("sha256", secret(publicKeyOrSecret)).update(bytes).digest(), signature);
}
/** Requires host-owned policy. A returned envelope does not authorize a tool call. */
export function verifyEnvelope(input: Envelope, policy: VerificationPolicy): Envelope {
  const e = envelopeFromObject(input), s = e.signature;
  if (!record(policy) || !Array.isArray(policy.keys) || !Array.isArray(policy.allowedAttributes) || !record(policy.context)) throw new PspError("INVALID_POLICY");
  validateJson(policy.context);
  if (Object.values(policy.context).some(v => typeof v !== "string") || policy.allowedAttributes.some(a => typeof a !== "string")) throw new PspError("INVALID_POLICY");
  if(policy.keys.some(k=>!record(k)||typeof k.id!=="string")) throw new PspError("INVALID_POLICY");
  const matches = policy.keys.filter(k => k.id === (s.kid ?? s.secretId));
  if (matches.length !== 1) throw new PspError(matches.length ? "INVALID_POLICY" : "UNKNOWN_KEY");
  const key = matches[0]!;
  if (key.status !== "active") throw new PspError("REVOKED_KEY");
  if (key.algorithm !== s.algorithm) throw new PspError("KEY_ALGORITHM_MISMATCH");
  if (!verifySignature(e, key.material)) throw new PspError("INVALID_SIGNATURE");
  const time = validateTime(s.timestamp, s.expires, policy.now, Object.hasOwn(policy,"clockSkew")?policy.clockSkew!:0);
  if (time !== "valid") throw new PspError(time === "invalid" ? "INVALID_TIME" : time.toUpperCase());
  if (!Array.isArray(key.trustLevels) || key.trustLevels.some((v:unknown)=>typeof v!=="number"||!Number.isInteger(v)||v<0||v>5) || !Array.isArray(key.sectionTypes) || key.sectionTypes.some((v:unknown)=>typeof v!=="string") || !record(key.scope) || typeof key.allowUnscoped !== "boolean") throw new PspError("INVALID_POLICY");
  if (!key.trustLevels.includes(s.trustLevel ?? 2) || !key.sectionTypes.includes(s.sectionType)) throw new PspError("UNAUTHORIZED_KEY");
  const attrs = s.attributes ?? {};
  if (Object.keys(attrs).some(k => !policy.allowedAttributes.includes(k))) throw new PspError("UNSUPPORTED_ATTRIBUTE");
  for (const [k,v] of Object.entries(key.scope)) if (typeof v !== "string" || policy.context[k] !== v) throw new PspError("SCOPE_MISMATCH");
  for (const [k,v] of Object.entries(policy.context)) {
    if (Object.hasOwn(attrs, k) ? attrs[k] !== v : !key.allowUnscoped) throw new PspError("SCOPE_MISMATCH");
  }
  if (!key.allowUnscoped && !Object.keys(policy.context).length) throw new PspError("SCOPE_MISMATCH");
  return e;
}

File: implementations/python/packages/core/src/psp_cdl_core/crypto.py
# SPDX-License-Identifier: Apache-2.0
"""Cryptographic adapter; no network discovery or implicit trust store."""
import hashlib
import hmac
from typing import TypedDict, NotRequired

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey, Ed25519PublicKey

from .json_codec import PspError, validate_json
from .signatures import decode_signature, encode_signature, envelope_from_object, signature_input, validate_time, integer
from .markup import Document, document_to_object


class TrustedKey(TypedDict):
    id: str
    algorithm: str
    material: bytes
    status: str
    trustLevels: list[int]
    sectionTypes: list[str]
    scope: dict[str, str]
    allowUnscoped: bool


class VerificationPolicy(TypedDict):
    keys: list[TrustedKey]
    now: float
    context: dict[str, str]
    allowedAttributes: list[str]
    clockSkew: NotRequired[int]


def _key(raw: bytes, length: int | None = None) -> bytes:
    if type(raw) is not bytes or (len(raw) != length if length else len(raw) < 32):
        raise PspError("INVALID_KEY")
    return raw


def sign_envelope(data: object, metadata: dict, private_key: bytes) -> dict:
    value = encode_signature(bytes(64 if metadata.get("algorithm") == "ed25519" else 32))
    e = envelope_from_object({"signature": {**metadata, "value": value}, "data": data})
    s = e["signature"]
    if validate_time(s["timestamp"], s["expires"], s["timestamp"]) != "valid":
        raise PspError("INVALID_TIME")
    message = signature_input(e)
    output = Ed25519PrivateKey.from_private_bytes(_key(private_key, 32)).sign(message) if s["algorithm"] == "ed25519" else hmac.digest(_key(private_key), message, hashlib.sha256)
    s["value"] = encode_signature(output)
    return e


def verify_signature(value: dict, public_key_or_secret: bytes) -> bool:
    """Crypto only; verify_envelope additionally enforces time, scope and key authority."""
    e = envelope_from_object(value)
    s = e["signature"]
    signature = decode_signature(s["value"], s["algorithm"])
    message = signature_input(e)
    if s["algorithm"] == "hmac-sha256":
        return hmac.compare_digest(signature, hmac.digest(_key(public_key_or_secret), message, hashlib.sha256))
    key = Ed25519PublicKey.from_public_bytes(_key(public_key_or_secret, 32))
    try:
        key.verify(signature, message)
        return True
    except InvalidSignature:
        return False


def sign_document(document: Document, metadata: dict, private_key: bytes, preserve_source: bool = True) -> dict:
    return sign_envelope(document_to_object(document, preserve_source), {**metadata, "contentType": "json"}, private_key)


def verify_envelope(value: dict, policy: VerificationPolicy) -> dict:
    """Host-owned policy is required. Returning an envelope is not tool authorization."""
    e = envelope_from_object(value)
    s = e["signature"]
    if type(policy) is not dict or type(policy.get("keys")) is not list or type(policy.get("allowedAttributes")) is not list or type(policy.get("context")) is not dict:
        raise PspError("INVALID_POLICY")
    validate_json(policy["context"])
    if any(type(v) is not str for v in policy["context"].values()) or any(type(v) is not str for v in policy["allowedAttributes"]):
        raise PspError("INVALID_POLICY")
    if any(type(k) is not dict or type(k.get("id")) is not str for k in policy["keys"]):
        raise PspError("INVALID_POLICY")
    matches = [k for k in policy["keys"] if k.get("id") == s.get("kid", s.get("secretId"))]
    if len(matches) != 1:
        raise PspError("INVALID_POLICY" if matches else "UNKNOWN_KEY")
    key = matches[0]
    if key.get("status") != "active":
        raise PspError("REVOKED_KEY")
    if key.get("algorithm") != s["algorithm"]:
        raise PspError("KEY_ALGORITHM_MISMATCH")
    if not verify_signature(e, key.get("material")):
        raise PspError("INVALID_SIGNATURE")
    temporal = validate_time(s["timestamp"], s["expires"], policy.get("now"), policy.get("clockSkew", 0))
    if temporal != "valid":
        raise PspError("INVALID_TIME" if temporal == "invalid" else temporal.upper())
    if type(key.get("trustLevels")) is not list or any(not integer(v, 0, 5) for v in key["trustLevels"]) or type(key.get("sectionTypes")) is not list or any(type(v) is not str for v in key["sectionTypes"]) or type(key.get("scope")) is not dict or type(key.get("allowUnscoped")) is not bool:
        raise PspError("INVALID_POLICY")
    if s.get("trustLevel", 2) not in key["trustLevels"] or s["sectionType"] not in key["sectionTypes"]:
        raise PspError("UNAUTHORIZED_KEY")
    attrs = s.get("attributes", {})
    if set(attrs) - set(policy["allowedAttributes"]):
        raise PspError("UNSUPPORTED_ATTRIBUTE")
    if any(type(v) is not str or policy["context"].get(k) != v for k, v in key["scope"].items()):
        raise PspError("SCOPE_MISMATCH")
    for k, v in policy["context"].items():
        if (attrs[k] != v) if k in attrs else not key["allowUnscoped"]:
            raise PspError("SCOPE_MISMATCH")
    if not key["allowUnscoped"] and not policy["context"]:
        raise PspError("SCOPE_MISMATCH")
    return e
